File: nodejs/examples/ProducerPriorityMessageExample.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Producer with Priority Message Example
 * 
 * This example demonstrates how to send priority messages that will be
 * consumed based on their priority level. Higher priority messages are
 * delivered before lower priority ones.
 * 
 * Key points for priority messages:
 * - Priority must be >= 0
 * - Higher priority messages are delivered first
 * - Cannot be used with messageGroup, deliveryTimestamp, or liteTopic
 * - Transactional messages cannot have priority set
 */
import { Producer } from '../src';
import { endpoints, sessionCredentials, namespace } from './ProducerSingleton';

async function main() {
  console.log('========== Producer Priority Message Example ==========');

  // Create producer instance
  const producer = new Producer({
    namespace,
    endpoints,
    sessionCredentials,
    maxAttempts: 3,
  });

  try {
    await producer.startup();
    console.log('Producer started successfully!\n');

    // Send messages with different priority levels
    console.log('Sending priority messages...');
    
    // Low priority message (priority = 1)
    console.log('\n1. Sending LOW priority message (priority=1)...');
    const lowPriorityReceipt = await producer.send({
      topic: 'yourPriorityTopic',
      tag: 'low-priority',
      body: Buffer.from(JSON.stringify({
        type: 'LOW_PRIORITY',
        content: 'This is a low priority notification',
        timestamp: Date.now(),
      })),
      priority: 1, // Lower number = lower priority
    });
    console.log('   ✓ Low priority message sent:', lowPriorityReceipt.messageId);

    // Medium priority message (priority = 5)
    console.log('\n2. Sending MEDIUM priority message (priority=5)...');
    const mediumPriorityReceipt = await producer.send({
      topic: 'yourPriorityTopic',
      tag: 'medium-priority',
      body: Buffer.from(JSON.stringify({
        type: 'MEDIUM_PRIORITY',
        content: 'This is a medium priority alert',
        timestamp: Date.now(),
      })),
      priority: 5, // Medium priority
    });
    console.log('   ✓ Medium priority message sent:', mediumPriorityReceipt.messageId);

    // High priority message (priority = 10)
    console.log('\n3. Sending HIGH priority message (priority=10)...');
    const highPriorityReceipt = await producer.send({
      topic: 'yourPriorityTopic',
      tag: 'high-priority',
      body: Buffer.from(JSON.stringify({
        type: 'HIGH_PRIORITY',
        content: 'This is a high priority urgent alert',
        timestamp: Date.now(),
      })),
      priority: 10, // Higher number = higher priority
    });
    console.log('   ✓ High priority message sent:', highPriorityReceipt.messageId);

    // Critical priority message (priority = 100)
    console.log('\n4. Sending CRITICAL priority message (priority=100)...');
    const criticalReceipt = await producer.send({
      topic: 'yourPriorityTopic',
      tag: 'critical-priority',
      body: Buffer.from(JSON.stringify({
        type: 'CRITICAL_PRIORITY',
        content: 'This is a critical emergency notification',
        timestamp: Date.now(),
      })),
      priority: 100, // Very high priority
    });
    console.log('   ✓ Critical priority message sent:', criticalReceipt.messageId);

    console.log('\n✓ All priority messages sent successfully!');
    console.log('\nNote: Messages will be delivered to consumers based on priority.');
    console.log('Higher priority messages (e.g., priority=100) will be delivered');
    console.log('before lower priority ones (e.g., priority=1).\n');

    // Demonstrate mutual exclusivity constraints
    console.log('Demonstrating mutual exclusivity constraints...\n');
    
    try {
      // This will fail: priority + messageGroup
      await producer.send({
        topic: 'yourPriorityTopic',
        body: Buffer.from('test'),
        priority: 5,
        messageGroup: 'group-1',
      });
      console.log('✗ ERROR: Should have failed when setting both priority and messageGroup');
    } catch (error: any) {
      console.log('✓ Correctly rejected: priority + messageGroup');
      console.log(`  Error: ${error.message}\n`);
    }

    try {
      // This will fail: priority + deliveryTimestamp
      await producer.send({
        topic: 'yourPriorityTopic',
        body: Buffer.from('test'),
        priority: 5,
        deliveryTimestamp: new Date(Date.now() + 60000),
      });
      console.log('✗ ERROR: Should have failed when setting both priority and deliveryTimestamp');
    } catch (error: any) {
      console.log('✓ Correctly rejected: priority + deliveryTimestamp');
      console.log(`  Error: ${error.message}\n`);
    }

    try {
      // This will fail: negative priority
      await producer.send({
        topic: 'yourPriorityTopic',
        body: Buffer.from('test'),
        priority: -1,
      });
      console.log('✗ ERROR: Should have failed with negative priority');
    } catch (error: any) {
      console.log('✓ Correctly rejected: negative priority');
      console.log(`  Error: ${error.message}\n`);
    }

    console.log('All validation checks passed!\n');

  } catch (error) {
    console.error('Failed to send priority messages:', error);
    throw error;
  } finally {
    await producer.shutdown();
  }
}

// Run example
main().catch(console.error);

File: csharp/examples/ProducerPriorityMessageExample.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Org.Apache.Rocketmq;

namespace examples
{
    /// <summary>
    /// Demonstrates how to send priority messages using Apache RocketMQ C# client.
    /// Priority messages allow you to assign different priority levels to messages,
    /// where higher priority messages are consumed before lower priority ones.
    /// 
    /// Key Points:
    /// - Priority must be >= 0 (higher value = higher priority)
    /// - Cannot be used with: messageGroup, deliveryTimestamp, or liteTopic
    /// - Consumer Group for testing: GID-priority-consumer
    /// </summary>
    internal static class ProducerPriorityMessageExample
    {
        private static readonly ILogger Logger = MqLogManager.CreateLogger(typeof(ProducerPriorityMessageExample).FullName);

        private static readonly string AccessKey = Environment.GetEnvironmentVariable("ROCKETMQ_ACCESS_KEY");
        private static readonly string SecretKey = Environment.GetEnvironmentVariable("ROCKETMQ_SECRET_KEY");
        private static readonly string Endpoint = Environment.GetEnvironmentVariable("ROCKETMQ_ENDPOINT");

        internal static async Task QuickStart()
        {
            // Enable the switch if you use .NET Core 3.1 and want to disable TLS/SSL.
            // AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);

            // Configure client for local testing (no authentication required)
            var clientConfig = new ClientConfig.Builder()
                .SetEndpoints("127.0.0.1:8081")
                .Build();

            const string topic = "topic-priority";
            
            // Create producer with singleton pattern (recommended)
            var producer = await new Producer.Builder()
                .SetTopics(topic)  // Prefetch topic route for better performance
                .SetClientConfig(clientConfig)
                .Build();

            // Define message content
            var body = Encoding.UTF8.GetBytes("This is a priority message for testing");
            const string tag = "PriorityTest";
            
            Logger.LogInformation($"Consumer Group: GID-priority-consumer");
            Logger.LogInformation("Sending messages with different priority levels (higher value = higher priority)...");
            
            // Send messages with priority levels from low to high
            var priorities = new[] { 1, 3, 5, 8, 10 };

            foreach (var priority in priorities)
            {
                var message = new Message.Builder()
                    .SetTopic(topic)
                    .SetBody(body)
                    .SetTag(tag)
                    .SetKeys($"priority-{priority}")
                    .SetPriority(priority)  // Higher priority messages are consumed first
                    .Build();

                try
                {
                    var sendReceipt = await producer.Send(message);
                    Logger.LogInformation($"Sent priority message successfully, messageId={sendReceipt.MessageId}, priority={priority}");
                }
                catch (Exception ex)
                {
                    Logger.LogError(ex, $"Failed to send priority message, priority={priority}");
                }
            }

            Logger.LogInformation("\nAll priority messages sent. Messages will be consumed in priority order (highest first).");

            // Close the producer if you don't need it anymore.
            await producer.DisposeAsync();
        }
    }
}
